Methods to apply exhaustion on your own

xZise opened the pull request:
I added two methods to player to handle the hunger stuff more smooth. For example if you want that users have to “pay” for a teleportation with hunger it is easily possible with both methods.

`Player.getTotalExhaustion()` gets the total exhaustion of a player. For example if the total exhaustion is 100 the player could walk 10 km and will then have no hunger level/saturation level left.

`Player.exhaust(float)` exhausts a specific value from the player. This method is very useful if you need to exhaust values larger than 4. Instead of changing the `exhaustion` field directly this will apply instantly. For example if you set exhaustion to 8 it will need two ticks to apply completely, because it will reduce the exhaustion only by 4 in each tick. `Player.exhaust(float)` instead will recalculate all values with one call and already on the next tick the player won't be healed/starve.

You could exhaust at maximum the value of getTotalExhaustion() but you will instantly starve.

Both could be done by outside of CraftBukkit, but then it is impossible to change the hunger behavior without breaking the plugins.

The corresponding CB pull request: https://github.com/Bukkit/CraftBukkit/pull/498

Fabian

xZise committed:
+exhaustion methods; +food level change event for all changes; +util class;

Dinnerbone commented:
For the record: https://github.com/Bukkit/CraftBukkit/pull/498 is the equivalent craftbukkit pull. (Please include this in the description in future)

xZise commented:
I updated everything so it is now acting instantly but won't change the exhaustion level (only if needed). So you could add an exhaustion of 20 (which will long for 5 ticks) with the normal `Player.setExhaustion(...)` call and then call Player.exhaust(20) which won't change the previous set exhaustion (of a minimum of 20).

And it is now firing only one event with the total change (old value is still available through the getters of the player). And I added a Util class which only has a static method `between` which checks if a value is in between two other values. I use them to check if the foodlevel/saturation level aren't to high.

Last but not least I call set*() in the constructor to prevent that the previously defined value is outside the valid border.

Fabian

Dinnerbone commented:
I'm currently cleaning old Pull Requests, and this one will be closed without being accepted.

I'm very sorry for the state that our Pull Requests have come to, and I hope you can all forgive us for letting them sit for so long. We're so undermanned for this, it's unbelievable. We hope that our new Bleeding system will allow for newer pull requests to be picked up and handled by much more people than just ourselves, which will speed up things tremendously.

If you wish to reopen your pull request, please make sure that it's 100% up to date with the current head, and feel free to resubmit whereupon one of our team will properly review it and give it the attention it deserves.

Again, I'm very sorry for letting these pulls sit for so long. Thanks for your patience.

xZise commented:
Okay, I opened a [leaky issue](https://bukkit.atlassian.net/browse/BUKKIT-822) and updated the source code to the current head.

Fabian
